### PR TITLE
BPE support for SwitchBoard & Transformer config

### PR DESCRIPTION
**Pull request by Emrys365**
add BPE support in SwitchBoard and configurations for RNN and Transformer

### CODE REVIEW

**Comment by sw005320** [comment]
- Can you also make a RESULT file?
- We usually have only one `run.sh` to avoid confusions. If the BPE version is better, we can replace it with the main `run.sh`.

**Comment by sw005320** [comment]
Also, please try the transformer.

**Comment by Emrys365** [comment]
> * Can you also make a RESULT file?
> * We usually have only one `run.sh` to avoid confusions. If the BPE version is better, we can replace it with the main `run.sh`.

Okay, I will update the code structure and `RESULT` when the experiments are finished (about two days).

File: egs/swbd/asr1/RESULTS [hunk 1]
```diff
@@ -4,4 +4,20 @@ exp/train_nodup_a01_pt_blstmp_e6/decode_eval2000_beam20_eacc.best_p0.1_len0.0-0.
 exp/train_nodup_a01_pt_blstmp_e6/decode_rt03_beam20_eacc.best_p0.1_len0.0-0.0/result.txt:|  SPKR         | # Snt    # Wrd  |  Corr     Sub      Del      Ins     Err    S.Err  |
 exp/train_nodup_a01_pt_blstmp_e6/decode_rt03_beam20_eacc.best_p0.1_len0.0-0.0/result.txt:|  Sum/Avg      | 8422    321545  |  78.2    10.6     11.2     12.3    34.1     79.6  |
 exp/train_nodup_a01_pt_blstmp_e6/decode_train_dev_beam20_eacc.best_p0.1_len0.0-0.0/result.txt:|  SPKR       |  # Snt     # Wrd  |  Corr       Sub      Del       Ins      Err     S.Err  |
-exp/train_nodup_a01_pt_blstmp_e6/decode_train_dev_beam20_eacc.best_p0.1_len0.0-0.0/result.txt:|  Sum/Avg    |  3999     235886  |  87.4       6.4      6.2       5.0     17.6      65.4  |
\ No newline at end of file
+exp/train_nodup_a01_pt_blstmp_e6/decode_train_dev_beam20_eacc.best_p0.1_len0.0-0.0/result.txt:|  Sum/Avg    |  3999     235886  |  87.4       6.4      6.2       5.0     17.6      65.4  |
+
+$ grep -e Avg -e SPKR -m 2 exp/train_nodup_pytorch_transformer_e12_BPE500_lr5_30epoch/decode_*_decode_pytorch_transformer_ctcweight0.3/scoring/hyp.*.sys
```
Comment (sw005320) [inline egs/swbd/asr1/RESULTS]: We usually perform the transformer with 50-100 epochs.
Could you resume to train your model with such epochs?

File: egs/swbd/asr1/conf/tuning/decode_pytorch_transformer.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+lm-weight: 1.0
```
Comment (sw005320) [inline egs/swbd/asr1/conf/tuning/decode_pytorch_transformer.yaml]: Did you tune the lm weight?
This seems to be too large.

File: egs/swbd/asr1/conf/tuning/decode_rnn.yaml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+lm-weight: 1.0
```
Comment (sw005320) [inline egs/swbd/asr1/conf/tuning/decode_rnn.yaml]: Did you tune the lm weight?
This seems to be too large.

File: egs/swbd/asr1/conf/tuning/train_pytorch_transformer.yaml [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+maxlen-in: 512  # if input length  > maxlen-in, batchsize is automatically reduced
+maxlen-out: 150 # if output length > maxlen-out, batchsize is automatically reduced
+
+# optimization related
+sortagrad: 0 # Feed samples from shortest to longest ; -1: enabled for all epochs, 0: disabled, other: enabled for 'other' epochs
+opt: noam
+accum-grad: 2
+grad-clip: 5
+patience: 0
+epochs: 30
```
Comment (sw005320) [inline egs/swbd/asr1/conf/tuning/train_pytorch_transformer.yaml]: As I mentioned, you can increase it (say 50). If the validation score is still improved, you could increase it more.

File: egs/swbd/asr1/conf/tuning/train_pytorch_transformer.yaml [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+dropout-rate: 0.1
+
+# transformer specific setting
+backend: pytorch
+model-module: "espnet.nets.pytorch_backend.e2e_asr_transformer:E2E"
+transformer-input-layer: conv2d     # encoder architecture type
+transformer-lr: 10.0
```
Comment (sw005320) [inline egs/swbd/asr1/conf/tuning/train_pytorch_transformer.yaml]: I saw that you used 5.0 in the RESULTS. Please update it with your tuned one.

File: egs/swbd/asr1/conf/train.yaml [hunk 6]
```diff
@@ -1,20 +1,20 @@
 # network architecture
```
Comment (sw005320) [inline egs/swbd/asr1/conf/train.yaml]: Please use the symlink instead of the actual files. See the other recipes.

File: egs/swbd/asr1/run.sh [hunk 7]
```diff
@@ -195,7 +265,12 @@ if [ ${stage} -le 4 ] && [ ${stop_stage} -ge 4 ]; then
         feat_recog_dir=${dumpdir}/${rtask}/delta${do_delta}
 
         # split data
-        splitjson.py --parts ${nj} ${feat_recog_dir}/data.json
+        if ${use_bpe}; then
```
Comment (sw005320) [inline egs/swbd/asr1/run.sh]: same comment.
Please use BPE as default.

**Comment by sw005320** [comment]
Please also fix the shell check in https://travis-ci.org/espnet/espnet/jobs/552603882#L1301-L1308

File: egs/swbd/asr1/RESULTS (hunk 1 above)
Comment (Emrys365) [reply egs/swbd/asr1/RESULTS]: Okay, I have already tried using 30 epochs and 100 epochs.

| Model | WER/S.Err (overall) | WER/S.Err (callhm) | WER/S.Err (swbd) |
| ---- | ---- | ---- | ---- |
| Transformer (15 epochs) | **20.4/59.3** | **26.0/63.3** | 14.7/53.5 |
| Transformer (30 epochs) | 20.0/59.4 | 26.0/64.5 | **14.0/52.0** |
| Transformer (100 epochs) | 20.3/59.8 | 26.5/65.4 | 14.1/51.6 |
| RNN | 22.0/60.1 | 28.5/64.6 | 15.6/53.5 |

File: egs/swbd/asr1/conf/tuning/decode_pytorch_transformer.yaml (hunk 2 above)
Comment (Emrys365) [reply egs/swbd/asr1/conf/tuning/decode_pytorch_transformer.yaml]: No, I didn't tune lm-weight. I can try with different lm-weights now.

**Comment by Emrys365** [comment]
> As I mentioned several places, Please do not arrange a recipe in your own way, which makes it difficult to maintain them.
> You should try to follow an existing recipe as much as possible.

Ok, I'll use BPE as default and rearrange the code.

**Comment by Emrys365** [comment]
> Please also fix the shell check in https://travis-ci.org/espnet/espnet/jobs/552603882#L1301-L1308

Okay, The second issue is fixed.
And the first one is due to a  glob range expression in the command, but it is actually used as one plain string, which also appears in the transcription of the training data (`data/train/text`).

File: egs/swbd/asr1/run.sh [hunk 8]
```diff
@@ -87,74 +96,135 @@ if [ ${stage} -le 1 ] && [ ${stop_stage} -ge 1 ]; then
...
+            data2json.sh --feat ${feat_recog_dir}/feats.scp \
+                --nlsyms ${nlsyms} ${data_root}/${rtask} ${dict} > ${feat_recog_dir}/data.json
+        done
+        echo "Stage 2 (char) finish!"
+    fi
 fi
 
 if [ -z ${tag} ]; then
     expname=${train_set}_${backend}_$(basename ${train_config%.*})
+    if ${use_bpe}; then
```
[the remark this answers is not in the document]
Comment (sw005320) [reply egs/swbd/asr1/run.sh]: Please remove this flag. You don't have `${use_bpe}`, right?

File: egs/swbd/asr1/run.sh [hunk 9]
```diff
@@ -180,8 +250,8 @@ if [ ${stage} -le 3 ] && [ ${stop_stage} -ge 3 ]; then
         --minibatches ${N} \
         --verbose ${verbose} \
         --resume ${resume} \
-        --train-json ${feat_tr_dir}/data.json \
```
[the remark this answers is not in the document]
Comment (sw005320) [reply egs/swbd/asr1/run.sh]: Again, please follow the way that the other recipe is doing. See
https://github.com/espnet/espnet/blob/master/egs/librispeech/asr1/run.sh#L222-L223

**Comment by sw005320** [comment]
@Emrys365, could you please follow the other BPE-based recipe (e.g., https://github.com/espnet/espnet/blob/master/egs/librispeech/asr1/run.sh) as much as possible?
If you arrange it in your own way, it is hard for us to maintain it.

**Comment by Emrys365** [comment]
> @Emrys365, could you please follow the other BPE-based recipe (e.g., https://github.com/espnet/espnet/blob/master/egs/librispeech/asr1/run.sh) as much as possible?
> If you arrange it in your own way, it is hard for us to maintain it.

Sorry, I'll rearrange the code.

**Comment by sw005320** [comment]
Please fix https://travis-ci.org/espnet/espnet/jobs/559855949#L1390-L1392

**Comment by sw005320** [comment]
Thanks a lot!

**Pull request merged**